File: packages/session-engine/src/risk.ts
// packages/session-engine/src/risk.ts — Client-side risk calculation

const BASE_TIMEOUT = 120; // seconds

export interface RiskContext {
    biometricScore: number;        // 0–100
    deviceTrustScore: number;      // 0–80
    isGeoAnomaly: boolean;
    isTrustedDevice: boolean;
    isOffHours: boolean;           // outside 09:00–19:00
    maxActiveTxnAmount: number;    // in paise
    failedTotpLastHour: number;
    isInGeoFence: boolean;
}

export interface RiskResult {
    adaptedTimeout: number;        // seconds
    riskLevel: "LOW" | "MEDIUM" | "HIGH";
    activeFactors: Array<{ label: string; impact: "+" | "-"; delta: number }>;
    requiresStepUp: boolean;
}

export function calculateRisk(ctx: RiskContext): RiskResult {
    let delta = 0;
    const factors: RiskResult["activeFactors"] = [];
    let requiresStepUp = false;

    // High-value transaction shortens timeout
    if (ctx.maxActiveTxnAmount > 40_000_00) { // > ₹4,00,000
        delta -= 30;
        factors.push({ label: "Active TXN > ₹4L", impact: "+", delta: 30 });
    }

    // Trusted device relaxes timeout
    if (ctx.isTrustedDevice) {
        delta += 45;
        factors.push({ label: "Trusted device", impact: "-", delta: 45 });
    }

    // Off-hours tightens timeout
    if (ctx.isOffHours) {
        delta -= 10;
        factors.push({ label: "Off-hours access", impact: "+", delta: 10 });
    }

    // In geo-fence relaxes timeout
    if (ctx.isInGeoFence) {
        delta += 20;
        factors.push({ label: "Within geo-fence", impact: "-", delta: 20 });
    }

    // Geo anomaly shortens significantly
    if (ctx.isGeoAnomaly) {
        delta -= 45;
        requiresStepUp = true;
        factors.push({ label: "Geo anomaly detected", impact: "+", delta: 45 });
    }

    // Biometric confidence
    if (ctx.biometricScore >= 85) {
        delta += 30;
        factors.push({ label: "High biometric confidence", impact: "-", delta: 30 });
    } else if (ctx.biometricScore < 65) {
        delta -= 20;
        factors.push({ label: "Low biometric score", impact: "+", delta: 20 });
        if (ctx.biometricScore < 40) {
            requiresStepUp = true;
        }
    }

    // Recent TOTP failures
    if (ctx.failedTotpLastHour >= 3) {
        delta -= 45;
        requiresStepUp = true;
        factors.push({ label: "Multiple TOTP failures", impact: "+", delta: 45 });
    }

    const adaptedTimeout = Math.max(45, Math.min(14400, BASE_TIMEOUT + delta)); // clamp 45s – 4h

    const riskScore = -delta;
    const riskLevel: RiskResult["riskLevel"] =
        riskScore <= 0 ? "LOW" : riskScore <= 40 ? "MEDIUM" : "HIGH";

    return {
        adaptedTimeout,
        riskLevel,
        activeFactors: factors,
        requiresStepUp,
    };
}

File: apps/api/routers/risk.py
from fastapi import APIRouter, Depends, Request
from services.risk_engine import calculate_adapted_timeout, RiskContext

router = APIRouter(prefix="/api/risk", tags=["risk"])

@router.post("/score")
async def compute_risk_score(request: Request):
    """Compute risk score from a provided context payload"""
    body = await request.json()
    ctx = RiskContext(
        user_id=body.get("user_id", "unknown"),
        session_id=body.get("session_id", "unknown"),
        biometric_score=body.get("biometric_score", 100),
        device_trust_score=body.get("device_trust_score", 50),
        is_geo_anomaly=body.get("is_geo_anomaly", False),
        is_trusted_device=body.get("is_trusted_device", True),
        is_off_hours=body.get("is_off_hours", False),
        max_active_txn_amount=body.get("max_active_txn_amount", 0),
        failed_totp_last_hour=body.get("failed_totp_last_hour", 0),
        is_in_geo_fence=body.get("is_in_geo_fence", True),
    )
    result = calculate_adapted_timeout(ctx)
    return {
        "adapted_timeout": result.adapted_timeout,
        "risk_level": result.risk_level,
        "active_factors": result.active_factors,
        "requires_step_up": result.requires_step_up,
    }
